### PR TITLE
WIP TypeScript definitions

### PR DESCRIPTION
**Pull request by ctaggart**
This is just reopening #613 that @madroneropaulo put together in August for #608. I need TypeScript definitions. This is work-in-progress. Feel free to contribute. Here are ways you can contribute:

- [ ] fill in all the [layer styles](https://github.com/mapbox/react-native-mapbox-gl/pull/1137#pullrequestreview-113039802)
- [ ] map [the constants](https://github.com/mapbox/react-native-mapbox-gl/pull/1137#pullrequestreview-113040259)

### CODE REVIEW

**Comment by nitaliano** [comment]
I think we would be better off opening the PR here https://github.com/DefinitelyTyped/DefinitelyTyped once we have the types ready

**Comment by ctaggart** [comment]
@nitaliano That works too. I'm fine with either way. Shall we work on them here until they are ready?

**Comment by nitaliano** [comment]
yup that's fine by me

**Comment by henrikra** [comment]
I think it would be better idea to add types to this repo since if/when the api changes it is required to changes TS types too. This way they are not lagging behind like every other types on DefinitelyTyped

**Comment by edwardgalligan** [comment]
I'd second @henrikra's suggestion here. DefinitelyTyped is typically designed for supporting projects that don't maintain their own official definitions, rather than hosting the definitions of those projects that do.

I'm away until mid-April but will look at contributing to this PR if/when/as soon as I can manage.

File: package.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
   "version": "6.1.0",
   "author": "Bobby Sudekum",
   "main": "./javascript/index.js",
+  "types": "./index.d.ts",
```
Comment (henrikra) [inline package.json]: You don't need to define this when your type file is in the project root and it is named `index.d.ts` 👍 So you can remove this line

**Comment by henrikra** [comment]
Good job @ctaggart !

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,594 @@
...
+interface LineCapStyle {
+    lineCap?: "butt" | "round" | "square"
+}
+
+interface LineJoinStyle {
+    lineJoin?: "bevel" | "round" | "miter"
+}
+
+interface LineMiterLimitStyle {
+    // TODO
```
Comment (ctaggart) [inline index.d.ts:118]: I would be wonderful if someone could fill in all the layer styles

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+    /** Source press listener, gets called when a user presses one of the children layers only
+if that layer has a higher z-index than another source layers */
+    onPress?: () => void
+
+    // hitbox: Shape // TODO
+}
+
+// https://github.com/mapbox/react-native-mapbox-gl/blob/master/docs/ShapeSource.md
+interface ShapeSourceProps {
+    // TODO
```
Comment (ctaggart) [inline index.d.ts:314]: fill out `ShareSourceProps`

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+}
+
+// https://github.com/mapbox/react-native-mapbox-gl/blob/master/docs/RasterSource.md
+interface RasterSourceProps {
+    // TODO
```
Comment (ctaggart) [inline index.d.ts:319]: fill out `RasterSourceProps`

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+
+    class RasterLayer extends Component<RasterLayerProps> {}
+
+    class SymbolLayer extends Component<SymbolLayerProps> {}
+
+    class Annotation extends Component {}
+
+    // Constants
+    // https://github.com/mapbox/react-native-mapbox-gl/blob/master/android/rctmgl/src/main/java/com/mapbox/rctmgl/modules/RCTMGLModule.java#L240-L270
+    const StyleURL: any // TODO
```
Comment (ctaggart) [inline index.d.ts:534]: There are a whole lot of constants that can be mapped like those below

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,594 @@
...
+    function initializeOfflinePacks(): Promise<void>;
+    function addOfflinePack(options: OfflinePackOptions, callback: () => void): Promise<void>;
+    function getOfflinePacks(callback: (packs: Progress []) => void): Promise<Progress []>;
+    function removeOfflinePack(callback: () => void): Promise<void>;
+    function addOfflinePackProgressListener(handler: (progress: Progress) => void): EmitterSubscription;
+    function addOfflineMaxAllowedTilesListener(handler: () => void): EmitterSubscription;
+    function addOfflineErrorListener(handler: () => void): EmitterSubscription;
+    function setOfflinePackProgressThrottleInterval(milis: number): void;
+
+    // TODO undocumented
```
Comment (ctaggart) [inline index.d.ts:584]: Where are the exported functions for the module documented?

File: tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,17 @@
+{
+  "compilerOptions": {
+    "target": "es5",
```
Comment (ctaggart) [inline tsconfig.json:3]: target es2015 instead?

**Comment by Aleksion** [comment]
What happened to this? Looking at the general progress on the repo it could seem like the project has gone stagnant?

**Comment by ctaggart** [comment]
@Aleksion I switch jobs and am no longer interested in pursuing this.

**Comment by henrikra** [comment]
@ctaggart Are you farmer nowadays? 👨‍🌾

**Pull request closed without merging**